### PR TITLE
Update politeness v2

### LINKED ISSUE #363: Update Politeness V2 based on revised package
# We need to update politeness_v2 and politeness_v2_helper with the revised package code.

I spoke with the developer of the software at a conference, and it seems that they've released revised code --- it outputs the same number of features as before, but the way that they calculate the features is now different. Let's align our code with the updated version to ensure that things now match.

Link to revised package: https://github.com/bbevis/politenessPy

### PR DESCRIPTION
**Pull request by sundy1994**
fix #363

### CODE REVIEW

File: src/team_comm_tools/features/politeness_v2_helper.py [hunk 1]
```diff
@@ -218,29 +218,85 @@ def bare_command(doc):
...
+                wh_count += 1
+                counted_sentences.add(sent.start)
+                break
+            # Yes/No pattern
+            if t1_lower in yesno_aux and t2_lower in pronoun_followers:
+                yesno_count += 1
+                counted_sentences.add(sent.start)
+                break
+    return yesno_count, wh_count
+    # sentences = [str(sent) for sent in doc.sents if '?' in str(sent)]
```
Comment (xehu) [inline src/team_comm_tools/features/politeness_v2_helper.py]: This seems like the old logic commented out, which we can remove once we establish the new logic works!

File: src/team_comm_tools/features/keywords.py [hunk 2]
```diff
@@ -7365,7 +7393,7 @@
         "Gratitude": [
             " thank ",
             " thanks ",
-            " thank you ",
+            #" thank you ",
```
Comment (xehu) [inline src/team_comm_tools/features/keywords.py]: Is it correct that we're commenting this out because otherwise we'd double count "thank you very much" ?

File: src/team_comm_tools/features/politeness_v2_helper.py [hunk 3]
```diff
@@ -218,29 +218,85 @@ def bare_command(doc):
...
+        'which': {'is', 'are', 'was', 'can', 'should'}
+    }
+    # Auxiliaries that typically initiate Yes/No questions
+    yesno_aux = {'do', 'does', 'did', 'have', 'has', 'had',
+                #  'can', 'could', 'will', 'would', 
+                 'may', 'might', 'shall', 'should',
+                 'is', 'are', 'was', 'were', 'am'}
+    # Pronouns that often follow auxiliaries in Yes/No questions
+    pronoun_followers = {'i', 'you', 'we', 'he', 'she', 'they', 'it'}
+    # filler_words = {'ok', 'so', 'well', 'like', 'you know', 'i mean', 'actually', 'basically', 'right', 'just', 'uh', 'um', 'oh', 'hmm', 'like'}
```
Comment (xehu) [inline src/team_comm_tools/features/politeness_v2_helper.py]: Are we no longer ignoring filler words like 'ok so' ?

File: src/team_comm_tools/features/politeness_v2_helper.py [hunk 4]
```diff
@@ -218,29 +218,85 @@ def bare_command(doc):
...
+    
+    wh_count = 0
+    yesno_count = 0
+    counted_sentences = set()
+    for sent in doc.sents:
+        sent_text = sent.text.strip()
+        sent_tokens = list(sent)
+        if not sent_tokens:
+            continue
+        # Method 1: Find question sentences by checking for '?' at end
```
Comment (xehu) [inline src/team_comm_tools/features/politeness_v2_helper.py]: So there's a bug with this logic, which is that, if the sentence doesn't end with a question mark, the fallback method is very error prone. For example:

Question(nlp("can you tell me what is your name?")) yields YesNo = 1, WH = 0 (correct)
Question(nlp("can you tell me what is your name")) yields YesNo = 0, WH = 1 (incorrect)

Why can't we use logic that looks for the question words, but then applies the logic with the search_tags, which helps to get around some of these problems?

**Review by xehu — CHANGES_REQUESTED** [review]
See error in the logic flow in the inline comments! I think we might need to make a small adjustment.

File: src/team_comm_tools/features/keywords.py (hunk 2 above)
Comment (sundy1994) [reply src/team_comm_tools/features/keywords.py]: Yes. Both "thank" and "thank you" will be counted. We discussed this issue with Burint and he made the changes.

File: src/team_comm_tools/features/politeness_v2_helper.py (hunk 3 above)
Comment (sundy1994) [reply src/team_comm_tools/features/politeness_v2_helper.py]: Yes. This was my own solution to this test case: "so Which part should we do first?". Burint updated his code later, so I adapted that instead.

File: src/team_comm_tools/features/politeness_v2_helper.py [hunk 5]
```diff
@@ -218,29 +218,79 @@ def bare_command(doc):
...
-            n += 1
-
-    return all_qs - n, n
+    # POS tags for WH-words like who/what/where
+    search_tags = {'WRB', 'WP', 'WDT'}
+    # WH-words and common auxiliaries that follow them in real questions
+    wh_words = {'what', 'who', 'where', 'when', 'why', 'how', 'which'}
+    wh_followers = {
+        'what': {'are', 'is', 'do', 'does', 'can', 'should', 'might'},
+        'who': {'am', 'is', 'are', 'was', 'can', 'should'},
```
Comment (xehu) [inline src/team_comm_tools/features/politeness_v2_helper.py]: @sundy1994 I found a bug when testing this. It turns out that if you do not add 'am' to WH_followers, things like _who am I_ or _what am I supposed to do_ are not detected as WH_questions. But there are still some bugs with this... I'll follow up on Slack.

Comment (xehu) [reply src/team_comm_tools/features/politeness_v2_helper.py]: UPDATE - I realized that it actually doesn't make much sense to separate the `wh_followers` from the other auxiliaries, so I refactored this so that we only use a consistent set of auxiliaries

**Comment by xehu** [comment]
[politeness_test.ipynb](https://github.com/user-attachments/files/24547423/politeness_test.ipynb)

I wrote a much more robust set of tests as I tried to iterate and figure out edge cases to the question function. Basically, I ended up just trying to come up with test cases, fixing them, and eventually I ended up reworking a good chunk of the function in order to catch all of the errors.

I think this is probably the best version of the question detector we have so far, and perhaps we can (1) update the test cases with these expected values and (2) make a PR or comment to the other repository with a note on the suggested fixes. Ultimately, I think it's better that we have a more accurate version of the question extraction, than to follow the source material if it is error-prone.

File: tests/data/cleaned_data/test_chat_level.csv [hunk 6]
```diff
@@ -88,14 +88,38 @@ I respond to that too",num_block_quote_responses,2
 1,A,hello,Hello_receptiveness_yeomans,1
 1,B,So how should we answer this,Token_count_receptiveness_yeomans,6
 1,A,We can start here. What is the question?,YesNo_Questions_receptiveness_yeomans,0
-1,B,I am not sure. Where is the rest of our team?,WH_Questions_receptiveness_yeomans,1
 1,B,"Please help me figure this out, I really want to do well on this please",Please_receptiveness_yeomans,2
+1,B,I am not sure. Where is the rest of our team?,WH_Questions_receptiveness_yeomans,1
+1,B,I am not sure. Where is the rest of our team?,First_Person_Single_receptiveness_yeomans,1
+1,B,"Well please help me figure this out, I really want to do well on this please okay",factuality_politeness_convokit,1
+1B,A,is where different from why?,YesNo_Questions_receptiveness_yeomans,1
```
Comment (xehu) [inline tests/data/cleaned_data/test_chat_level.csv:95]: @sundy1994 I added in this more robust set of question tests to test all the edge cases that the code (hopefully) now catches!

**Review by xehu — APPROVED** [review]
Our version of the politeness questions is more robust than ever, and we also have updated tests to confirm that we get the edge cases right. I say we merge this in!
